File: src/main/java/ar/edu/unq/spring/service/impl/PersonajeServiceImpl.java
package ar.edu.unq.spring.service.impl;

import ar.edu.unq.spring.modelo.Personaje;
import ar.edu.unq.spring.modelo.exception.NombreDePersonajeRepetido;
import ar.edu.unq.spring.persistence.repository.PersonajeRepository;
import ar.edu.unq.spring.service.interfaces.PersonajeService;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.NoSuchElementException;
import java.util.Set;

@Service
@Transactional
public class PersonajeServiceImpl implements PersonajeService {

    private final PersonajeRepository personajeRepository;
    public PersonajeServiceImpl(PersonajeRepository personajeRepository) {
        this.personajeRepository = personajeRepository;
    }

    @Override
    public Set<Personaje> allPersonajes() {
        return Set.copyOf(personajeRepository.findAll());
    }

    @Override
    public void guardarPersonaje(Personaje personaje) {
        try {
            personajeRepository.save(personaje);
        } catch (DataIntegrityViolationException e) {
            throw new NombreDePersonajeRepetido(personaje.getNombre());
        }
    }

    @Override
    public Personaje recuperarPersonaje(Long personajeId) {
        return personajeRepository.findById(personajeId).orElseThrow(() -> new NoSuchElementException("Personaje not found with id: " + personajeId));
    }

    @Override
    public void clearAll() {
        personajeRepository.deleteAll();
    }
}

File: src/main/kotlin/ar/edu/unq/spring/service/impl/PersonajeServiceImpl.kt
package ar.edu.unq.spring.service.impl

import ar.edu.unq.spring.modelo.Personaje
import ar.edu.unq.spring.modelo.exception.NombreDePersonajeRepetido
import ar.edu.unq.spring.persistence.PersonajeDAO
import ar.edu.unq.spring.service.interfaces.PersonajeService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.dao.DataIntegrityViolationException
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.PlatformTransactionManager
import org.springframework.transaction.TransactionDefinition
import org.springframework.transaction.TransactionStatus
import org.springframework.transaction.support.DefaultTransactionDefinition

@Service
class PersonajeServiceImpl() : PersonajeService {

    @Autowired private lateinit var personajeDAO: PersonajeDAO
    @Autowired private lateinit var transactionManager: PlatformTransactionManager

    override fun allPersonajes(): Collection<Personaje> {
        return personajeDAO.findAll().toMutableSet()
    }

    override fun guardarPersonaje(personaje: Personaje) {
        val definition = DefaultTransactionDefinition()
        definition.propagationBehavior = TransactionDefinition.PROPAGATION_REQUIRED
        val transactionStatus: TransactionStatus = transactionManager.getTransaction(definition)
        try {
            personajeDAO.save(personaje)
            transactionManager.commit(transactionStatus)
        } catch (e: DataIntegrityViolationException) {
            transactionManager.rollback(transactionStatus)
            throw NombreDePersonajeRepetido(personaje.nombre!!)
        } catch (ex: Exception) {
            transactionManager.rollback(transactionStatus)
            throw ex
        }
    }

    override fun recuperarPersonaje(personajeId: Long): Personaje? {
        return personajeDAO.findByIdOrNull(personajeId)
    }
}
